test-runtime/build.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of subxt.
+//
+// subxt is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// subxt is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with subxt.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{
+    env,
+    fs,
+    net::TcpListener,
+    path::Path,
+    process::Command,
+    sync::atomic::{
+        AtomicU16,
+        Ordering,
+    },
+    thread,
+    time,
+};
+
+static SUBSTRATE_BIN_ENV_VAR: &str = "SUBSTRATE_NODE_PATH";
+
+#[async_std::main]
+async fn main() {
+    // Select substrate binary to run based on env var.
+    let substrate_bin =
+        env::var(SUBSTRATE_BIN_ENV_VAR).unwrap_or_else(|_| "substrate".to_owned());
+
+    // Run binary.
+    let port = next_open_port()
+        .expect("Cannot spawn substrate: no available ports in the given port range");
+    let cmd = Command::new(&substrate_bin)
+        .arg("--dev")
+        .arg("--tmp")
+        .arg(format!("--rpc-port={}", port))
+        .spawn();
+    let mut cmd = match cmd {
+        Ok(cmd) => cmd,
+        Err(e) => {
+            panic!("Cannot spawn substrate command '{}': {}", substrate_bin, e)
+        }
+    };
+
+    // Download metadata from binary; retry until successful, or a limit is hit.
+    let metadata_bytes: sp_core::Bytes = {
+        const MAX_RETRIES: usize = 20;
+        let mut retries = 0;
+        let mut wait_secs = 1;
+        loop {
+            if retries >= MAX_RETRIES {
+                panic!("Cannot connect to substrate node after {} retries", retries);
+            }
+            let res =
+                subxt::RpcClient::try_from_url(&format!("http://localhost:{}", port))
+                    .await
+                    .unwrap()
```

Comment:
This might panic if it can't connect

Reply:
Ah yes, I think I took it to be if the url was malformed, and was surprised to then have to move it after the await..

Reply:
Actually, I did have a perusal through the code, and it's all OK as is (if the url is WS, it tries to establish a connection (hence the await), but if it's HTTP it won't try to connect and will only error if not an http(s) URL). 

I'll change to an expect to document this!

Reply:
yeah, I think we should kill `http` anyway because it won't work for subscriptions.